File: ruby/1.rb
#!/usr/bin/env ruby

require 'timing_method'

# This is woefully inefficient.
#
# def adding_function
#   puts (1...1000).select{|i|
#     i %3 == 0 || i %5 == 0
#   }.reduce(:+)
# end
#
# timing_method do
#   adding_function
# end

#
# Let’s look at the details of our function and take as example n=3.
# We would have to add:
# 3+6+9+12+......+999=3*(1+2+3+4+...+333)
# For n=5 we would get:
# 5+10+15+...+995=5*(1+2+....+199)
# Now note that 199=995/5 but also 999/5 rounded down to the nearest integer.
# In many programming languages there exists a separate operator for that: div or \.
# If we now also note that 1+2+3+...+p=½*p*(p+1)  our program becomes:
#
max = ARGV[2].to_i - 1
val1 = ARGV[0].to_i
val2 = ARGV[1].to_i

def SumDivisibleByN(n, max)
  p = max / n
  (n*(p*(p+1)))/ 2
end

# http://en.wikipedia.org/wiki/Euclidean_algorithm
def gcd(a, b)
  while b != 0
    t = b
    b = (a%b)
    a = t
  end
  a
end

# lcm = product of x*y divided by the gcd of x & y
def lcm(x, y)
  (x*y) / gcd(x, y)
end

# This is crazy fast.
timing_method do
text = <<CUT
Add all the natural numbers below maximum number that are multiples of val1 or val2.
CUT
  if max >= 0
    puts SumDivisibleByN(val1, max) + SumDivisibleByN(val2, max) - SumDivisibleByN(lcm(val1, val2), max)
    # The last part of the previous equation should be the least common multiple
    # Int.lcm works only in ruby 1.9
    # puts SumDivisibleByN(val1, max) + SumDivisibleByN(val2, max) - SumDivisibleByN(val1.lcm(val2), max)
  else
    puts text
    print "usage:  #{$0} <first divisor> <second divisor> <maximum number>\n"
    exit
  end
end

File: c/1.c
/*
 * =====================================================================================
 *
 *       Filename:  1.c
 *
 *    Description:  Project Euler Problem 1
 *
 *        Version:  1.0
 *        Created:  08/13/2011 07:01:27
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Matt Simpson (Maui Threv), mr.threv@gmail.com
 *        Company:
 *
 * =====================================================================================
 */

#include <stdio.h>
#include <math.h>

/*
 * ===  FUNCTION  ======================================================================
 *         Name:  gcd
 *  Description:  Greatest Common Divisor
 * =====================================================================================
 */
int gcd (int x, int y )
{
  int t = 0;
  while ( y != 0 ) {
    t = y;
    y = (x%y);
    x = t;
  }
  return x;
}		/* -----  end of function gcd  ----- */


/*
 * ===  FUNCTION  ======================================================================
 *         Name:  lcm
 *  Description:  Least Common Multiple
 * =====================================================================================
 */
int lcm ( int x, int y )
{
  return (x * y) / gcd(x, y);
}		/* -----  end of function lcm  ----- */


/*
 * ===  FUNCTION  ======================================================================
 *         Name:  sumdivisiblebyn
 *  Description:  Returns the number of values up to max that are divisible by n.
 * =====================================================================================
 */
int sumdivisiblebyn ( int n, int max)
{
  int p = max / n;
  return n*(p*(p+1))/2;
}		/* -----  end of function sumdivisiblebyn  ----- */


/* If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
 * Find the sum of all the multiples of 3 or 5 below 1000.
 */
int main(int argc, const char *argv[])
{
  int x = 3;
  int y = 5;
  int xy = lcm(3, 5);
  int max = 999;

  int sum = sumdivisiblebyn(x, max) + sumdivisiblebyn(y, max) - sumdivisiblebyn(xy, max);

  printf("The sum of all multiples of %d or %d that are below %d is %d.\n", x, y, max, sum);

  return 0;
}
